File: DynamicProgramming/139.word-break.ts
// @leet start
function wordBreak(s: string, wordDict: string[]): boolean {
  const choice = new Set<number>();
  const m = new Set<string>();
  for (const v of wordDict) {
    m.add(v);
    choice.add(v.length);
  }
  const dp = new Array(s.length + 1).fill(false);
  dp[0] = true;
  for (let i = 0; i <= s.length; i++) {
    for (const step of choice) {
      if (i >= step && dp[i - step]) {
        if (m.has(s.slice(i - step, i))) {
          dp[i] = true;
          break;
        }
      }
    }
  }

  return dp[s.length];
}

// @leet end

File: DynamicProgramming/139.word-break.py
from typing import List


# @leet start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [True] + [False] * len(s)
        choise = set([len(i) for i in wordDict])
        m = set(wordDict)
        for i in range(len(s) + 1):
            for step in choise:
                if i >= step and dp[i - step]:
                    if s[i - step : i] in m:
                        dp[i] = True
                        break

        return dp[len(s)]


# @leet end


class Solution2:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = {}

        def dfs(i: int) -> bool:
            if i == len(s):
                return True
            if i > len(s):
                return False
            if i in memo:
                return memo[i]
            ret = False
            for item in wordDict:
                if s[i : i + len(item)] == item:
                    if dfs(i + len(item)):
                        ret = True
                        break
            memo[i] = ret
            return ret

        return dfs(0)
